pkg/operator/controller/dns_status.go

```diff
@@ -0,0 +1,146 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// syncDNSStatus computes the current status of dns and
+// updates status upon any changes since last sync.
+func (r *reconciler) syncDNSStatus(dns *operatorv1.DNS, clusterIP, clusterDomain string, ds *appsv1.DaemonSet) error {
+	updated := dns.DeepCopy()
+	updated.Status.ClusterIP = clusterIP
+	updated.Status.ClusterDomain = clusterDomain
+	updated.Status.Conditions = computeDNSStatusConditions(updated.Status.Conditions, clusterIP, ds)
+	if !dnsStatusesEqual(updated.Status, dns.Status) {
+		if err := r.client.Status().Update(context.TODO(), updated); err != nil {
+			return fmt.Errorf("failed to update dns status: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// computeDNSStatusConditions computes dns status conditions
+// based on the status of ds.
```

Comment:
Technically, that should be "based on the status of ds and clusterIP"... but maybe I am being excessively fussy.